### PR TITLE
feat: add basic Juju storage support

### PR DESCRIPTION
**Pull request by marcoppenheimer**
##### `feat: configure Kafka log.dirs to use mounted storage volumes`
- Blocks and does not start if < minimum:range storages found

## Up Next
- More gracefully handle storage adding/removing/attach/detach events for currently running charm. Currently none handled
- Integration tests for all storage events
    - This is not added to this PR as to 'generate' logs to these mounted storage directories, we have to build a robust `producer` so that topic metadata is stored across these dirs
    - This hasn't been done yet, TBC
- NOTE - All future PR's will be merging to feature branch `feat/storage` to keep PRs more reviewable.

### CODE REVIEW

File: metadata.yaml [hunk 1]
```diff
@@ -26,3 +26,12 @@ requires:
...
     interface: kafka_client
+
+storage:
+  log-data:
+    type: filesystem
+    description: Directories where the log data is stored
+    minimum-size: 50M
+    location: /var/snap/kafka/common
+    multiple:
+      range: "12"
```
Comment (Mehdi-Bendriss) [inline metadata.yaml:37]: Why is this needed?

File: src/charm.py [hunk 2]
```diff
@@ -78,6 +84,19 @@ def unit_peer_data(self) -> MutableMapping[str, str]:
...
+    def _on_storage_attached(self, event: StorageAttachedEvent) -> None:
+        path = event.storage.location if event.storage else None
+        if not path:
+            logger.error("Unable to find storage in StorageAttachedEvent")
+            event.defer()
+            return
+
+        self.unit_peer_data.update({"logs": "attached"})
+
+    def _on_storage_detatched(self, _: StorageDetachingEvent) -> None:
```
Comment (Mehdi-Bendriss) [inline src/charm.py:96]: nit: `detached`, also is this callback registered?

**Review by Mehdi-Bendriss — APPROVED** [review]

File: src/charm.py (hunk 2 above)
Comment (marcoppenheimer) [reply src/charm.py:96]: Valid nit.
It's not, it'll be in future PRs. This isn't merging to `main`, just to a feature branch.

Comment (marcoppenheimer) [reply src/charm.py:96]: It's left out because I need to actually validate the behavior of Kafka when you rug-pull a directory from it.

File: metadata.yaml (hunk 1 above)
Comment (marcoppenheimer) [reply metadata.yaml:37]: The multiple?

Requirement from product. Idea is that it helps parallelize writes. Some recommendations online for a production system suggest having 12 dirs as a good baseline.

Comment (Mehdi-Bendriss) [reply metadata.yaml:37]: Thanks for explaining!

**Review by welpaolo — APPROVED** [review]

File: metadata.yaml (hunk 1 above)
Comment (deusebio) [reply metadata.yaml:39]: What if we put 1-12? I suppose this would allow the charm to work in the "unsupported" case of having 1 storage, and then add storage up to 12 maximum slots with "add-storage"

File: metadata.yaml [hunk 3]
```diff
@@ -26,3 +28,12 @@ requires:
 provides:
   kafka-client:
     interface: kafka_client
+
+storage:
+  log-data:
+    type: filesystem
+    description: Directories where the log data is stored
+    minimum-size: 50M
```
Comment (deusebio) [inline metadata.yaml:36]: The specs was kinda asking for 1TB... 

Is there any way we could set this? Should setting be done in bundles maybe?

File: src/charm.py [hunk 4]
```diff
@@ -78,6 +84,19 @@ def unit_peer_data(self) -> MutableMapping[str, str]:
...
         return self.peer_relation.data[self.unit]
 
+    def _on_storage_attached(self, event: StorageAttachedEvent) -> None:
+        path = event.storage.location if event.storage else None
+        if not path:
+            logger.error("Unable to find storage in StorageAttachedEvent")
+            event.defer()
+            return
+
+        self.unit_peer_data.update({"logs": "attached"})
```
Comment (deusebio) [inline src/charm.py:94]: Do this trigger a new rolling restart of Kafka ? I suppose as we add storage during operations, we should trigger here a restart to provide the added directories to the properties file, right?

File: metadata.yaml (hunk 3 above)
Comment (marcoppenheimer) [reply metadata.yaml:36]: That will be addressed in future PRs. This is for basic functionality at this stage.

File: src/charm.py (hunk 4 above)
Comment (marcoppenheimer) [reply src/charm.py:94]: That will be addressed in future PRs. This is for basic functionality at this stage.

File: metadata.yaml (hunk 1 above)
Comment (marcoppenheimer) [reply metadata.yaml:39]: We can address that in later PRs once we have a clearer idea of how the workload should handle such events.
